### PR TITLE
gridded-temp-data

### PR DESCRIPTION
**Pull request by ateucher**
WIP. Downloading ECCC homogenized temperature data from https://open.canada.ca/data/en/dataset/d6813de6-b20a-46cc-8990-01862ae15c5f"﻿

### CODE REVIEW

**Comment by ateucher** [comment]
@boshek @BonnieJRobert this is ready for review.

I've converted it to a package structure, so you can run `devtools::install_deps()` to ensure all the packages you need are installed, and use `devtools::load_all()` to load the packages for debugging etc.

The `_targets.R` file can be your guide for what needs review. I would suggest running `tar_make()` to start, it should complete but will take a while. Let me know if the `future` parallel setup works for you.

Basically any targets that get built in the final pipeline should be reviewed - and I think all of the functions for review are in `R/fn-heatwave.R`. You can ignore anything in the `scratch` folder.

Also, the two csv files generated in the `out/heatwave_summaries` folder should be looked at as well, as they are for use in the next step.

Please let me know if you don't understand (or disagree with) any of the logic.

Thank you!

**Comment by BonnieJRobert** [comment]
required raacq and weathercan to be downloaded separately from GitHub

File: DESCRIPTION [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+           comment = c(ORCID = "0000-0002-9270-7884")),
+    person("Province of British Columbia", role = "cph")
+  )
+Description: What the package does (one paragraph).
+Imports:
+    targets,
+    tarchetypes,
+    arrow,
+    dplyr,
+    weathercan,
```
Comment (BonnieJRobert) [inline DESCRIPTION:23]: remove weathercan and rcaaqs

File: R/fn-heatwave.R [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+    janitor::clean_names() |>
+    dplyr::mutate(from = paste(from, x6, sep = "-"),
+                  to = paste(to, x8, sep = "-")) |>
+    dplyr::select(-x6, -x8)
+
+  sf::st_as_sf(stations, coords = c("long_deg", "lat_deg"), crs = 4326)
+}
+
+get_bc_target_stations <- function(stations, buffer, crs) {
+  bc_buff <- sf::st_buffer(st_union(bcmaps::bc_bound()), dist = buffer * 1000)
```
Comment (BonnieJRobert) [inline R/fn-heatwave.R:177]: tar_make() error:
Error: callr subprocess failed: could not find function "st_union"

**Review by boshek — APPROVED** [review]
This looks great @ateucher. Very easy to understand the flow of the pipeline. I am really liking how a targets pipeline enables a fairly easy review. Most of the comments are of the "optional" nature. The only one that I think needs some consideration is whether this should be generalized beyond LHAs to accommodate any future geographies.

File: _targets.R [hunk 3]
```diff
@@ -16,30 +16,126 @@ library(targets)
...
-  tar_target(area_of_interest, health_lha() %>% filter(grepl("Vancouver", LOCAL_HLTH_AREA_NAME, ignore.case = TRUE) |
-                                                       LOCAL_HLTH_AREA_NAME %in% c("Greater Nanaimo", "Kamloops", "Smithers", "Nelson", "Central Okanagan", "Greater Victoria"))),
-  #tar_target(pm25_data, pm25(area_of_interest, start_date = start_date, end_date = end_date)), ##pm data has some issues with arrow col specs
-  tar_target(weather_data, weather(area_of_interest, start_date = start_date, end_date = end_date, normals = FALSE, ask = FALSE)),
-  tar_target(area_burned_over_time, calc_area_burned_over_time(area_of_interest)),
-  tar_target(flood_example, hy_daily_flows("08NN002", start_date = start_date, end_date = end_date))
+  tar_target(area_of_interest,
+             health_lha() %>%
+               filter(LOCAL_HLTH_AREA_NAME %in% LHAs) %>%
+               st_transform(st_crs(dem))),
```
Comment (BonnieJRobert) [inline _targets.R:77]: consider explicit namespaces

File: R/fn-heatwave.R [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+temp_tps <- function(df) {
+  df <- na.omit(df)
+  indep <- df[, c("x", "y", "elevation")]
+
+  fields::Tps(x = indep, Y = df$temp, miles = FALSE)
+}
+
+interpolate_daily_temps <- function(model_list, dem, variable) {
+  stars_list <- future.apply::future_lapply(model_list, function(mod) {
+    predict(dem, mod)
```
Comment (BonnieJRobert) [inline R/fn-heatwave.R:231]: consider explicit namespaces

**Review by BonnieJRobert — APPROVED** [review]
very interesting!  Unable to get several of the targets (pixel_*, daily_*, target_stations and area_of_interest) running in debug mode so unfortunately  I was only able to review up to heatwave_targets.  Lot's of learning, thanks!

File: R/fn-heatwave.R [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+    dplyr::select(stn_id, x = X, y = Y, elevation = elev_m)
+
+  days <- unique(temp_data$date)
+
+  out <- future.apply::future_lapply(days, function(d) {
+    data <- temp_data[temp_data$date == d, , drop = FALSE]
+    df <- dplyr::left_join(stations, data, by = "stn_id") |>
+      dplyr::select(x, y, elevation, temp)
+    temp_tps(df)
+  }, future.seed = 13L)
```
[the remark this answers is not in the document]
Comment (ateucher) [reply R/fn-heatwave.R:217]: Good idea

File: R/fn-heatwave.R [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+#'
+#' @param stars_cube stars cube of tmax for timeseries of interest
+#' @param start_date date
+#' @param end_date date
+#'
+#' @return list of climatologies (outputs of heatwaveR::ts2clm);
+#' one element for each pixel
+generate_pixel_climatologies <- function(stars_cube, start_date, end_date) {
+  not_na_pixels <- !is.na(stars_cube$tmax)
+  num_pixels_per_slice <- apply(not_na_pixels, 3, sum)
```
[the remark this answers is not in the document]
Comment (ateucher) [reply R/fn-heatwave.R:268]: Correct - looping over time periods and checking each time period has the same number of non-na pixels. Added a comment to clarify

File: _targets.R [hunk 7]
```diff
@@ -16,30 +16,126 @@ library(targets)
...
+hw_output_dir <- "out/heatwave_summaries"
+dir.create(hw_output_dir, showWarnings = FALSE, recursive = TRUE)
 
 # Load --------------------------------------------------------------------
 # time variables
-time_vars <- list(
-  tar_target(start_date, as.Date('1981-01-01')),
-  tar_target(end_date, Sys.Date())
+static_vars <- list(
+  tar_target(start_date, as.Date('1990-04-01')),
```
[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:45]: That's right, happy to discuss more. Though this has made me realize that I think I should be generating a full year's worth of surfaces to pass into `generate_pixel_climatologies` (which calls `heatwaveR::ts2clm()`) so that it has the appropriate data at the beginning and end of the seasonal window.

File: R/fn-heatwave.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ateucher) [reply R/fn-heatwave.R:177]: Thanks! I've removed that line now

File: R/fn-heatwave.R [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+  lha_clim_summary |>
+    dplyr::mutate(thresh2 = temp_0.75 >= 30) |>
+    dplyr::select(LOCAL_HLTH_AREA_CODE,
+                  t, temp = temp_mean, seas = seas_mean,
+                  thresh = thresh_mean,
+                  thresh2) |>
+    (\(x) split(x, x$LOCAL_HLTH_AREA_CODE))() |>
+    lapply(\(x) {
+      heatwaveR::detect_event(x, threshClim2 = x$thresh2, minDuration = 2,
+                              categories = TRUE, climatology = TRUE, S = FALSE)
```
[the remark this answers is not in the document]
Comment (ateucher) [reply R/fn-heatwave.R:390]: done, thanks!

File: _targets.R [hunk 9]
```diff
@@ -16,30 +16,126 @@ library(targets)
...
+  tar_target(dem, get_dem(res = raster_res)),
+  tar_target(analysis_temps, arrow::open_dataset(ahccd_parquet_path) %>%
+               filter(date >= start_date, date <= end_date,
+                      stn_id %in% target_stations$stn_id,
+                      measure %in% c("daily_max", "daily_min")) %>%
+               collect()),
+  tar_target(daily_tmax_models, model_temps_xyz(temp_data = filter(analysis_temps, measure == "daily_max"),
+                                           stations = target_stations,
+                                           months = 4:9)),
+  tar_target(daily_temps_stars_cube,
```
[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:105]: This is not an uncommon warning in using `fields::Tps()` and not necessarily a bad thing. I am looking into it...

[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:105]: Here is the one explanation I can find... https://stat.ethz.ch/pipermail/r-help/2002-May/021562.html

File: _targets.R (hunk 3 above)
Comment (ateucher) [reply _targets.R]: Done thanks!

File: R/fn-heatwave.R (hunk 4 above)
Comment (ateucher) [reply R/fn-heatwave.R:239]: We can't namespace `predict()` because it is an `S3` method (`predict.stars()`)

File: _targets.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:45]: Addressed in https://github.com/bcgov/climate-disturbances/pull/10/commits/53b0897e7f40c9a639bb24b57a9526f3c62787c2

File: R/fn-heatwave.R [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+  rast <- dplyr::slice(stars_cube, "time", 1)
+  as.data.frame(rast) |>
+    dplyr::select(-tmax) |>
+    dplyr::mutate(pixel_id = paste(x, y, sep = ";")) |>
+    sf::st_as_sf(coords = c("x", "y"), crs = st_crs(stars_cube)) |>
+    sf::st_join(
+      dplyr::select(area_of_interest,
+                    LOCAL_HLTH_AREA_CODE, LOCAL_HLTH_AREA_NAME)
+      ) |>
+    dplyr::filter(!is.na(LOCAL_HLTH_AREA_CODE))
```
[the remark this answers is not in the document]
Comment (ateucher) [reply R/fn-heatwave.R]: Addressed in 39d1269

File: _targets.R [hunk 11]
```diff
@@ -16,30 +16,126 @@ library(targets)
...
 
 #  climate data
 climate_targets <- list(
-  tar_target(area_of_interest, health_lha() %>% filter(grepl("Vancouver", LOCAL_HLTH_AREA_NAME, ignore.case = TRUE) |
-                                                       LOCAL_HLTH_AREA_NAME %in% c("Greater Nanaimo", "Kamloops", "Smithers", "Nelson", "Central Okanagan", "Greater Victoria"))),
-  #tar_target(pm25_data, pm25(area_of_interest, start_date = start_date, end_date = end_date)), ##pm data has some issues with arrow col specs
-  tar_target(weather_data, weather(area_of_interest, start_date = start_date, end_date = end_date, normals = FALSE, ask = FALSE)),
-  tar_target(area_burned_over_time, calc_area_burned_over_time(area_of_interest)),
-  tar_target(flood_example, hy_daily_flows("08NN002", start_date = start_date, end_date = end_date))
+  tar_target(area_of_interest,
```
[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:73]: Strange! I'm getting the same thing. I haven't figured out what the problem is, but I almost think it's a `targets` bug. Doing a trace on the error it never even gets to my code before it errors. I'll investigate further but I don't think it's a blocker to merge this - you?

[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:73]: Opened as a new issue #11

File: _targets.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (ateucher) [reply _targets.R:105]: Opened new issue #12
